### ISSUE TITLE
Type check data binding into element attributes

### ISSUE THREAD

**Issue by rictic** [Banno/polymer-rename#1]
You've likely already thought of this, if so excuse the enthusiastic brainstorming :) 

I think there's a huge additional win available by extending the type checking into data-binding between elements as well.

e.g. code like

```html
<dom-module id="foo-bar">
  <template>
    <baz-qux my-name="{{barName}}" count="[[barCount]]"></baz-qux>
  </template>
</dom-module>
```

could become something like

```javascript
(/** @this {FooBarElement} */ function() {
  var bazQuxElem = /** @type {BazQuxElement} */(document.createElement("baz-qux"));

  // Assign in both directions as this is a bidirectional binding.
  bazQuxElem.myName = this.barName;
  this.barName = bazQuxElem.myName;

  // Just one assignment expression because this is a one-way binding.
  bazQuxElem.count = this.barCount;

}).call(/** @type {FooBarElement} */ (document.createElement("foo-bar")))
```

**Comment by ChadKillingsworth** [comment]
Yes - the type system possibilities are huge. Even without this running it on my project started spitting out warnings like "Property 'foo' never defined on FooBarElement". That alone is a huge win - and one that gives polymer templating a strategic advantage over other popular Google frameworks.

My initial focus is to get a working tool because I have an internal deadline in mid-may that this is a prerequisite for.

Give me till at least the end of next week and I'll open this up to other contributors. Right now the amount of refactoring I'm doing would just create chaos.

**Comment by ChadKillingsworth** [comment]
Fixed by 18a6a7313216e0d295e21cafe41f6592839e80ed
